File: showSensorHexapod/src/specificworker.py
#
# Copyright (C) 2016 by YOUR NAME HERE
#
#    This file is part of RoboComp
#
#    RoboComp is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    RoboComp is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
#

import sys, os, Ice, traceback, time

from PySide import *
from genericworker import *

ROBOCOMP = ''
try:
	ROBOCOMP = os.environ['ROBOCOMP']
except:
	print '$ROBOCOMP environment variable not set, using the default value /opt/robocomp'
	ROBOCOMP = '/opt/robocomp'
if len(ROBOCOMP)<1:
	print 'genericworker.py: ROBOCOMP environment variable not set! Exiting.'
	sys.exit()


preStr = "-I"+ROBOCOMP+"/interfaces/ --all "+ROBOCOMP+"/interfaces/"
Ice.loadSlice(preStr+"JointMotor.ice")
from RoboCompJointMotor import *
Ice.loadSlice(preStr+"FootPreassureSensor.ice")
from RoboCompFootPreassureSensor import *



class SpecificWorker(GenericWorker):
    
        listaPatas = ["arm1","arm2","arm3","arm4","arm5","arm6"]
        
        listaMotores = ["arm1motor1","arm1motor2","arm1motor3",
                        "arm2motor1","arm2motor2","arm2motor3",
                        "arm3motor1","arm3motor2","arm3motor3",
                        "arm4motor1","arm4motor2","arm4motor3",
                        "arm5motor1","arm5motor2","arm5motor3",
                        "arm6motor1","arm6motor2","arm6motor3"]
        
        listaProgress = []
        lcdNumbers = []
        
        def __init__(self, proxy_map):
		super(SpecificWorker, self).__init__(proxy_map)
		self.timer.timeout.connect(self.compute)
		self.Period = 100
		self.timer.start(self.Period)
		
                self.lcdNumbers = [self.ui.lcdNumber_1,  self.ui.lcdNumber_3,  self.ui.lcdNumber_5, 
                                   self.ui.lcdNumber_2,  self.ui.lcdNumber_4,  self.ui.lcdNumber_6,
                                   self.ui.lcdNumber_7,  self.ui.lcdNumber_9,  self.ui.lcdNumber_11,
                                   self.ui.lcdNumber_8,  self.ui.lcdNumber_10, self.ui.lcdNumber_12,
                                   self.ui.lcdNumber_13, self.ui.lcdNumber_15, self.ui.lcdNumber_17,
                                   self.ui.lcdNumber_14, self.ui.lcdNumber_16, self.ui.lcdNumber_18]

	def setParams(self, params):
		#try:
		#	par = params["InnerModelPath"]
		#	innermodel_path=par.value
		#	innermodel = InnerModel(innermodel_path)
		#except:
		#	traceback.print_exc()
		#	print "Error reading config params"
		return True

	@QtCore.Slot()
	def compute(self):
		#print 'SpecificWorker.compute...'
		#try:
		#	self.differentialrobot_proxy.setSpeedBase(100, 0)
		#except Ice.Exception, e:
		#	traceback.print_exc()
		#	print e
		
		
                self.listaProgress = [self.ui.progressBar_1, self.ui.progressBar_2, self.ui.progressBar_3, self.ui.progressBar_4, self.ui.progressBar_5, self.ui.progressBar_6] 
                	
                for i in range(6):
                    sensor = self.footpreassuresensor_proxy.readSensor(self.listaPatas[i])
                    self.listaProgress[i].setValue(sensor*100/1023)

                for i in range (18):
                    try:
                        estado = self.jointmotor_proxy.getMotorState(self.listaMotores[i])
                        temperatura = estado.temperature
                        self.lcdNumbers[i].display(temperatura)
                    except Ice.Exception as e:
                        print e
                    
		return True

File: brainhexapod/src/specificworker.cpp
/*
 *    Copyright (C) 2016 by YOUR NAME HERE
 *
 *    This file is part of RoboComp
 *
 *    RoboComp is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    RoboComp is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "specificworker.h"

/**
* \brief Default constructor
*/
SpecificWorker::SpecificWorker(MapPrx& mprx) : GenericWorker(mprx)
{
	proxies[0]=legcontroller1_proxy;
	proxies[1]=legcontroller2_proxy;
	proxies[2]=legcontroller3_proxy;
	proxies[3]=legcontroller4_proxy;
	proxies[4]=legcontroller5_proxy;
	proxies[5]=legcontroller6_proxy;
	pasostate[0]="paso";
	pasostate[1]="empujar";
	pasostate[2]="empujar";
	pasostate[3]="paso";
	pasostate[4]="paso";
	pasostate[5]="empujar";
	osgView = new OsgView(this->frame);
	show();
	clkupdate.start(10);
	connect(&clkupdate, SIGNAL(timeout()), this, SLOT(updatevalues()));
}

/**
* \brief Default destructor
*/
SpecificWorker::~SpecificWorker()
{
	
}

bool SpecificWorker::setParams(RoboCompCommonBehavior::ParameterList params)
{
	string name = PROGRAM_NAME;
	qDebug()<<PROGRAM_NAME;
	base=QString::fromStdString(params[name+".base"].value);
	innerModel = new InnerModel(params[name+".InnerModel"].value);

	innerViewer = new InnerModelViewer(innerModel, "root", osgView->getRootGroup(), false);
	
	for(int i=1;i<=6;i++)
	{
		legs<<QString::fromStdString(params[name+".nameleg"+to_string(i)].value);
	}
	
	qDebug()<<"-----------------------legs----------------";
	qDebug()<<base;
	qDebug()<<legs;
	qDebug()<<"-----------------------legs----------------";
	
	for(int i=0;i<6;i++){
		statelegs[i] = proxies[i]->getStateLeg();
		legsp[i]=QVec::vec3(statelegs[i].x,statelegs[i].y,statelegs[i].z);
	}
	
	timer.start(Period);
	
	
	return true;
}


void SpecificWorker::compute()
{
// 	static int i = 0;
// 	if(i==6)
// 		i=0;
// 	bool menor=true;
// 	double incremento=0;
	if (allidel())
	{
// 		for(int i = 0; i<6; i++)
// 		{
// 			if(statelegs[i].y < -90)
// 			{
// 				menor=false;
// 				break;
// 			}
// 		}
// 		if(menor)
// 			for(int i = 0; i<6; i++)
// 				if(incremento < statelegs[i].y+96)
// 					incremento = statelegs[i].y+96;
// 		for(int i = 0; i<6; i++)
// 		{
// 			RoboCompLegController::PoseLeg p;
// 			p.ref =  statelegs[i].ref;
// 			p.vel = 50;
// 			p.x = statelegs[i].x;
// 			p.z = statelegs[i].z;
// 			p.y = statelegs[i].y-incremento;
// 			proxies[i]->setIKLeg(p,false);
// 		}
		for(int i = 0; i<6; i++)
		{
			qDebug()<<"envio a la pata "<< (i+1);
			proxies[i]->move(0,30,pasostate[i]);
			if(pasostate[i]=="paso")
				pasostate[i]="empujar";
			else
				pasostate[i]="paso";
		}
	}
	
}

bool SpecificWorker::allidel()
{
	bool noidel=true;
	QString idel;
	qDebug()<<"-----------------";
	for(int i=0; i<6; i++)
	{
		if(statelegs[i].idel==false)
		{
			noidel=false;
			qDebug()<<" Leg " << (i+1) << statelegs[i].idel;
		}
	}
	qDebug()<<"-----------------";
	return noidel;
}

void SpecificWorker::updatevalues()
{
	//update inner
	for(int i=0;i<6;i++)
		statelegs[i] = proxies[i]->getStateLeg();
	for(int i=0;i<6;i++)
	{
		RoboCompLegController::StateLeg s=statelegs[i];
		innerModel->updateJointValue(QString::fromStdString(s.q1.name),s.q1.pos);
		innerModel->updateJointValue(QString::fromStdString(s.q2.name),s.q2.pos);
		innerModel->updateJointValue(QString::fromStdString(s.q3.name),s.q3.pos);
	}
	innerViewer->update();
	osgView->autoResize();
	osgView->frame();
}






